src/query_plan/operation.rs

```diff
@@ -2776,6 +2965,49 @@ impl NormalizedFieldSelection {
     pub(crate) fn has_defer(&self) -> bool {
         self.field.has_defer() || self.selection_set.as_ref().is_some_and(|s| s.has_defer())
     }
+
+    fn can_add_to(
+        &self,
+        parent_type: &CompositeTypeDefinitionPosition,
+        schema: &ValidFederationSchema,
+    ) -> Result<bool, FederationError> {
+        match (parent_type, &self.field.data().field_position) {
```

Comment:
JS code actually checks whether we are pointing at the same object -> here we need to check the type names AND whether we point to the same schema (I did the same check in rebasing logic [here](https://github.com/apollographql/federation-next/blob/main/src/query_plan/operation.rs#L2717-L2722))